rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,697 @@
+| No                    | Single node migrations coordinated through a lease / lock |
+
+Idempotent migrations don't require coordination making the algorithm
+significantly less complex and will never require manual intervention to
+retry. We, therefore, prefer this solution, even though it introduces
+restrictions on migrations (4.2.1.1). For other alternatives that were
+considered see section [(5)](#5-alternatives).
+
+## 4.2.1 Idempotent migrations performed without coordination
+
+The migration system can be said to be idempotent if the same results are
+produced whether the migration was run once or multiple times. This property
+should hold even if new (up to date) writes occur in between migration runs
+which introduces the following restrictions:
+
+### 4.2.1.1 Restrictions
+
+1. All document transforms need to be deterministic, that is a document
+   transform will always return the same result for the same set of inputs. 
+2. It should always be possible to construct the exact set of inputs required
+   for (1) at any point during the migration process (before, during, after). 
+
+Although these restrictions require significant changes, it does not prevent
+known upcoming migrations such as [sharing saved-objects in multiple spaces](https://github.com/elastic/kibana/issues/27004) or [splitting a saved
+object into multiple child
+documents](https://github.com/elastic/kibana/issues/26602). To ensure that
+these migrations are idempotent, they will have to generate new saved object
+id's deterministically with e.g. UUIDv5.
+
+
+### 4.2.1.2 Migration algorithm: Cloned index per version
+Note:
+- The description below assumes the migration algorithm is released in 7.10.0.
+  So < 7.10.0 will use `.kibana` and >= 7.10.0 will use `.kibana_current`.
+- We refer to the alias and index that outdated nodes use as the source alias
+  and source index.
+- Every version performs a migration even if mappings or documents aren't outdated.
+
+1. Locate the source index by fetching aliases (including `.kibana` for versions prior to v7.10.0) `GET '/_alias/.kibana_current,.kibana_7.10.0,.kibana'`. The source index is the index the `.kibana_current` alias points to, or if it doesn’t exist, the index the `.kibana` alias points to.
+2. If `.kibana_current` and `.kibana_7.10.0` both exists and are pointing to the same index this version's migration has already been completed.
+   1. Because the same version can have plugins enabled at any point in time, perform the mappings update in step (6).
+   3. Skip to step (9) to start serving traffic.
+3. Fail the migration if:
+   1. `.kibana_current` is pointing to an index that belongs to a later version of Kibana .e.g. `.kibana_7.12.0_001`
+   2. The source index contains documents that belong to an unknown Saved Object type (from a disabled plugin). Log an error explaining that the plugin that created these documents needs to be enabled again or that these objects should be deleted. See section (4.2.1.4).
```

Comment:
I agree that this could be seen as a breaking change. But once a version upgrade has completed with unknown types it will introduce downtime and is quite complicated to fix (see 4.2.1.4.1 "perform a re-migration").

In contrast, failing to start a migration won't introduce downtime or difficult to recover from scenarios. 

To reduce any upgrade friction so that users don't get stuck on an old version because of this new behaviour we could introduce a runtime flag that will delete any unknown types from the target index. That means that using the flag will preserve the unknown types in the "backup" source index but these won't cause any problems like increasing the field count or users accidentally starting a migration without all the necessary plugins enabled. 

What do you think?

Reply:
I think it's going to be hard to argue for the originally proposed behavior to not be considered a breaking change. This situation doesn't only occur when a plugin is disabled, but also when a third-party plugin was uninstalled after creating a saved-object.

> To reduce any upgrade friction so that users don't get stuck on an old version because of this new behaviour we could introduce a runtime flag that will delete any unknown types from the target index. That means that using the flag will preserve the unknown types in the "backup" source index but these won't cause any problems like increasing the field count or users accidentally starting a migration without all the necessary plugins enabled.

I definitely think this would make it easier for a user to recover from this situation, but if the user has to manually specify the runtime flag, it'll still be considered a breaking change. I also don't think we should default to this behavior within 7.x because it could create a really poor user experience (which could be argued is a breaking change, but I don't think it's clear) if they were to perform the spaces example in section 4.2.1.4.

One potential solution which isn't mentioned in section 4.2.1.4 is to include a hash of all of the installed plugins in the name of the actual ES index. Ostensibly, this seems like it'd allow us to perform a migration whenever plugins are installed/uninstalled or enabled/disabled. Are there intricacies to this solution that I'm overlooking?

Reply:
Performing a migration on every enable/disable solves our problems, but the downside is it causes downtime for all nodes in the cluster. This shouldn't happen often and the downtime should usually be very minimal (probably only the time it takes to make a clone). But this could also be seen as a breaking change since currently a plugin could be enabled/disabled as a rolling configuration change while all nodes are online, but with this change the source index will be marked as read-only disabling writes from other nodes.

I'm also not immediately sure how we'll handle enabling, then disabling and enabling again on the same version. Either way I think it's worth adding this option to the RFC and fleshing it out.

Reply:
In 4.2.1.2 we have:
> 2. If `.kibana_current` and `.kibana_7.10.0` both exists and are pointing to the same index this version's migration has already been completed.
>    1. Because the same version can have plugins enabled at any point in time, perform the mappings update in step (6).
>    2. Skip to step (9) to start serving traffic.

I think we could perform step (7) which transforms outdated documents here too before starting to serve traffic and can't immediately think of any big downsides.

This still has the downside that we could run into future problems with incompatible mappings, but seems like a good solution for 7.x.

I've updated the RFC with these alternatives and new recommendation.

Reply:
> I think we could perform step (7) which transforms outdated documents here too before starting to serve traffic and can't immediately think of any big downsides.

This seems reasonable to me.